### ISSUE TITLE
Proposal - Selecting a subset of an object's properties

### ISSUE THREAD

**Issue by gregsdennis** [ietf-wg-jsonpath/draft-ietf-jsonpath-base#49]
There's no syntax for selecting multiple individual properties from a single object.  You either have to select a single property or the entire object.

I would like to propose a "union" or "subselection" syntax `(x,y,...)` where `x`, etc. are the properties to be returned.  The return value will be an object omitting all properties that are not included in the union declaration.

Using Goessner's example, I could use the path `$.store.book[*].(title,price)` to select only the title and price of all of the books.  It would return

```json
[
  { "title": "Sayings of the Century", "price": 8.95 },
  { "title": "Sword of Honour", "price": 12.99 },
  { "title": "Moby Dick", "price": 8.99 },
  { "title": "The Lord of the Rings", "price": 22.99 },
]
```

***NOTE** I understand the name conflicts with current usage of "union", but that be resolved by [changing that usage](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-jsonpath/issues/21).  I'm struggling to come up with another name.

**Comment by cburgmer** [comment]
I like to thing of there being at least two different problems many people try solving in this space: That of selecting values, and that of transforming the original JSON structure. I understand JSONPath to solve the former, but to be really bad at the latter (by choice).
Selecting a subset for me is clearly a transformation problem, as I can already use `$["a", "b"]` to select the necessary values.
So my personal stance is to not get into that.

**Comment by gregsdennis** [comment]
@cburgmer what's the support for something like `$["a","b"]`?  I'm not sure I've seen that anywhere.

**Comment by mkmik** [comment]
fwiw I'm also in the camp of "jsonpath selects and doesn't transforms"

**Comment by cburgmer** [comment]
> @cburgmer what's the support for something like `$["a","b"]`? I'm not sure I've seen that anywhere.

I'm talking about the union operator with two keys: https://cburgmer.github.io/json-path-comparison/results/union_with_keys.html. You will get the values, but not the keys. 

Java's com.jayway.jsonpath interestingly enough decides to return the object with the keys, contrary to the rest of the implementations (and Objective-C's SMJJSONPath which according to it's goal just follows com.jayway.jsonpath verbatim).

**Comment by danielaparker** [comment]
> Java's com.jayway.jsonpath interestingly enough decides to return the object with the keys, contrary to the rest of the 
> implementations (and Objective-C's SMJJSONPath which according to it's goal just follows com.jayway.jsonpath verbatim).

Jayway JsonPath has had this as an open issue since 2017, 

https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-jsonpath/issues/49#issuecomment-754812728

**Comment by danielaparker** [comment]
> I like to thing of there being at least two different problems many people try solving in this space: That of selecting values, and 
> that of transforming the original JSON structure. I understand JSONPath to solve the former, but to be really bad at the latter 
> (by choice). Selecting a subset for me is clearly a transformation problem, as I can already use `$["a", "b"]` to select the 
> necessary values. So my personal stance is to not get into that.

Agreed, and of course normalized path expressions cannot be produced when a query results in new JSON elements. Note however that many JSONPath implementations support functions such as `avg` or `sum` that do return results that are not in the original JSON instance.

**Comment by cburgmer** [comment]
For further reference, Ruby's jsonpath has a "parens notation" (https://cburgmer.github.io/json-path-comparison/results/parens_notation.html) to select a subset: `$(key,more)`. Other implementations might have found their own syntax for supporting something like that, this is the only one I came across though.

**Comment by goessner** [comment]
Interesting proposal regarding the syntax. 

I am also in the camp of "jsonpath selects and doesn't transform" (I've even been the first one :)

The example above could be rewritten to `$.store.book[?(@.title && @.price)]`, which returns a list of output pathes
```js
[ "$['store']['book'][0]",
  "$['store']['book'][1]",
  "$['store']['book'][2]",
  "$['store']['book'][3]"
]
```
that can be used then to do any post-processing ... here transform to property-reduced books.

No violation of the minimalism priciple this way!

Proposed syntax `(<expr1>,<expr2>,...)` might be used to separate a brackets child selector from that, what is still called 'union' at current.

**Comment by cabo** [comment]
See resolution of #109 at IETF 112.
As we don't do projection in -base, this is now closed/revisit-after-base-done.
